### PR TITLE
[SPARK-34369][SQL] Track number of pairs processed out of Join

### PR DESCRIPTION
**Pull request by sririshindra**
### What changes were proposed in this pull request?
Adding a new "number of matched pairs" metric to track number of pairs processed out of join.

### Why are the changes needed?
Often users face a scenario where even a modest skew in a join can lead to tasks appearing to be stuck, due to the O(n^2) nature of a join considering all pairs of rows with matching keys. When this happens users think that spark has gotten deadlocked. If there is a bound condition, the "number of output rows" metric may look typical. Other metrics may look very modest (eg: shuffle read). In those cases, it is very hard to understand what the problem is. There is no conclusive proof without getting a heap dump and looking at some internal data structures.

It would be much better if spark had a metric(which we propose be titled “number of matched pairs” as a companion to “number of output rows”) which showed the user how many pairs were being processed in the join. This would get updated in the live UI (when metrics get collected during heartbeats), so the user could easily see what was going on.

This would even help in cases where there was some other cause of a stuck executor (eg. network issues) just to disprove this theory. For example, you may have 100k records with the same key on each side of a join. That probably won't really show up as extreme skew in task input data. But it'll become 10B join pairs that spark works through, in one task. See  the Jira description for an example.

### Does this PR introduce _any_ user-facing change?
Yes, users will be able to see a new "number of matched pairs" metric whenever they look at a join node in the UI.

### How was this patch tested?
Added appropriate unit tests

### CODE REVIEW

**Comment by sririshindra** [comment]
Could you please take a look at this PR. cc: @maropu @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala [hunk 1]
```diff
@@ -57,7 +57,8 @@ case class BroadcastHashJoinExec(
   }
 
   override lazy val metrics = Map(
-    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+    "numMatchedPairs" -> SQLMetrics.createMetric(sparkContext, "number of matched pairs"))
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: `numMatchedPairs` -> `numJoinedRows`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 2]
```diff
@@ -89,13 +90,20 @@ case class BroadcastNestedLoopJoinExec(
     }
   }
 
-  @transient private lazy val boundCondition = {
+  private val numMatchedPairs = longMetric("numMatchedPairs")
+
+  @transient private lazy val boundCondition: InternalRow => Boolean =
     if (condition.isDefined) {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:96]: If `condition` is empty, `numOutputRows` == `numMatchedPairs `? If so, could we only show this new metric in the web UI when `condition` defined?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 3]
```diff
@@ -415,24 +431,26 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
...
+          BindReferences.bindReference(expr, streamedPlan.output ++ buildPlan.output).genCode(ctx)
+        val skipRow = s"${ev.isNull} || !${ev.value}"
+        s"""
+           |$eval
+           |${ev.code}
+           |if (!($skipRow))
+         """.stripMargin
+      } else {
+        ""
+      })
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:453]: Could you avoid unnecessary changes? For example;
```
    val checkCondition = if (condition.isDefined) {
      val expr = condition.get
      ...
    } else {
        ""
    }
    val some_val = s"$numMatched.add(1);\n$checkCondition"
    (matched, some_val, buildVars)
```

**Review by maropu — COMMENTED** [review]
Adding the new metric itself looks fine.

**Comment by maropu** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 2 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:96]: If the condition is empty, then `numOutputRows == numMatchedPairs` only in the case of InnerJoin. It is not necessarily the case for other types of join, because numMatchedPairs is only measuring the records when there is a common key on both the build side and the stream side.  For instance, in the case of leftOuterJoin `numOutputRows  = numMatchedPairs  + (number of non matched pairs on the left side of the join)`.
Moreover, if we define a metric and fail to populate it for some cases, then the existing unit tests in the SQLMetricsSuite will fail because of the way the unit tests fetch the actual metrics.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala (hunk 1 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: Sure, do you also want me to change the display name as well, `number of matched pairs` -> `number of joined rows`? I feel that `number of matched pairs` might be more precise and less confusing to the user. But if you think that  `number of joined rows` better I will change the name of the metric accordingly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 3 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Fixed in latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: >  I feel that number of matched pairs might be more precise and less confusing to the user.

Any reason? At first, I was not sure about what `pairs` means though.

Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: How about `number of matched rows` instead?  I wanted highlight the fact that this metric measures the number of keys matched between the stream side and the build side.  I also felt that users might not recognize the difference between `number of joined rows` and `number of output rows`. But `number of joined rows` is fine as well. If you think that users will understand that better I will change the name of the metric to `number of joined rows`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 2 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:96]: Ah, I see. What is the merit of being able to see this number for a user? I read the PR description and, at first, I thought this PR mainly targets at the case where a `condition` is given.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: sgtm

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 2 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:96]: - Yes, the metric is very useful especially when there is a condition. It helps the users to recognize skew. But even if there is no condition, it could help the users to see the number of matched rows. For instance, in case of outer joins users will be able to see just how many rows matched between the stream side and the build side without explicitly doing another inner join.   

- I also faced an implementation problem when I originally tried to show the metric only when there is a condition. If there is a metric defined, but if the metric is not used in some cases (For example if we don't increment the metric when there is no condition), then the tests in SQLMetricsSuite will fail because the accumulatorIds corresponding the metrics would get messed up. The size of metric values in `val metricValues = statusStore.executionMetrics(executionId)` and the metrics in `node.metrics` would become inconsistent. 

- It might also confuse the users to see the metric in some cases and not see the metric in other cases. It would probably be best to show the metric everytime there is a join node in the UI to avoid any confusion.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala (hunk 1 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala]: Updated the name of the metric to `number of matched rows` in the latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 4]
```diff
@@ -137,10 +144,19 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
...
+      (r: InternalRow) => {
+        numMatchedRows += 1
+        Predicate.create(condition.get, streamedPlan.output ++ buildPlan.output).eval(r)
+      }
+    } else {
+      (_: InternalRow) => {
+        numMatchedRows += 1
+        true
+      }
   }
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:160]: Indentation?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 5]
```diff
@@ -433,6 +450,7 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
     } else {
       ""
     }
+    val checkCondition = s"$numMatched.add(1);\n$conditionDef"
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:453]: Could you add `$numMatched.add(1);` directly before line 446 and line 451?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 6]
```diff
@@ -716,6 +741,7 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
     } else {
       s"$existsVar = true;"
     }
+    val checkCondition = s"$numMatched.add(1);\n$conditionDef"
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:744]: ditto. We had better add this before line 712 and line 717 directly.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 7]
```diff
@@ -277,20 +277,34 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+        val df2 = spark.sql(query2)
+        Seq(false, true).foreach { case  enableWholeStage =>
+          testSparkPlanMetrics(df2, 1, Map(
+            0L -> ("SortMergeJoin", Map(
+              "number of output rows" -> 3L,
+              "number of matched rows" -> 4L))),
+            enableWholeStage
+          )
+        }
+
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:307]: Shall we remove this empty line?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 8]
```diff
@@ -277,20 +277,34 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+            "number of matched rows" -> 4L))),
           nodeId2 -> (("Exchange", Map(
             "records read" -> 4L,
             "local blocks read" -> 2L,
             "remote blocks read" -> 0L,
             "shuffle records written" -> 2L)))),
           enableWholeStage
         )
+
+        val query2 = "SELECT * FROM testData2 JOIN testDataForJoin ON " +
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:296]: Why this is inside the `foreach` of line 281?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 9]
```diff
@@ -303,18 +317,35 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+        (query4, 8L, true),
+        (query4 + boundCondition1, 7L, false),
+        (query4 + boundCondition1, 7L, true),
+        (query1, 8L, true),
+        (query2, 8L, false),
+        (query2, 8L, true)).foreach { case (query, rows, enableWholeStage) =>
         val df = spark.sql(query)
         testSparkPlanMetrics(df, 1, Map(
           0L -> (("SortMergeJoin", Map(
             // It's 8 because we read 6 rows in the left and 2 row in the right one
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:346]: Please update this comments accordingly because this is invalid after this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 10]
```diff
@@ -324,13 +355,22 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+    ).foreach { case (boundCondition, dfLeft, dfRight, nodeId, rows, enableWholeStage) =>
+      var df = dfLeft.join(broadcast(dfRight), "key")
+      if (boundCondition) {
+        df = df.filter("value1 > value2")
+      }
       testSparkPlanMetrics(df, 2, Map(
         nodeId -> (("BroadcastHashJoin", Map(
-          "number of output rows" -> 2L)))),
+          "number of output rows" -> rows,
+          "number of matched rows" -> 2L)))),
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:373]: Is it always `2`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 11]
```diff
@@ -369,23 +409,38 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
-    Seq((0L, "right_outer", leftDf, rightDf, 10L, false),
-      (0L, "left_outer", rightDf, leftDf, 10L, false),
-      (1L, "right_outer", leftDf, rightDf, 10L, true),
-      (1L, "left_outer", rightDf, leftDf, 10L, true),
-      (2L, "left_anti", rightDf, leftDf, 8L, true),
-      (2L, "left_semi", rightDf, leftDf, 2L, true),
-      (1L, "left_anti", rightDf, leftDf, 8L, false),
-      (1L, "left_semi", rightDf, leftDf, 2L, false))
-      .foreach { case (nodeId, joinType, leftDf, rightDf, rows, enableWholeStage) =>
+  test("ShuffledHashJoin(left/right/outer/semi/anti, outer) metrics") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:412]: Please add `SPARK-34369: ` test prefix.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 4 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Done. Fixed in the latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 5 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Done. Fixed in the latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 6 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: @dongjoon-hyun I am slightly confused about this part. I added `$numMatched.add(1);` within the checkCondition itself. I did this by adding it directly  before line 742 and line 737 directly. That way every time checkCondition is invoked the metric is incremented. I hope that is what you are looking for. If not, could you please elaborate where I am making a mistake.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 7 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Done. Fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 8 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Yes, that was a mistake. Fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 9 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Yes, I removed the comment entirely as there are multiple queries and the comment is no longer valid for all the queries.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 10 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Done. Fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 11 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Done! fixed in latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 12]
```diff
@@ -48,6 +48,13 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
     s"$nodeName $joinType ${buildSide} ($opId)".trim
   }
 
+  /**
+   * When overriding metrics be sure to include numOutputRows & numMatchedRows
+   */
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+    "numMatchedRows" -> SQLMetrics.createMetric(sparkContext, "number of matched rows"))
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:56]: I feel it's kind of weird that we need to double define `metrics` here in `HashJoin`, and also its subclass `BroadcastHashJoinExec` and `ShuffledHashJoinExec`. It might introduce unexpected bug later. Could we pass the `numMatchedRows` to `HashJoin`, similar to what we did for `numOutputRows` before?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 13]
```diff
@@ -89,13 +90,20 @@ case class BroadcastNestedLoopJoinExec(
     }
   }
 
-  @transient private lazy val boundCondition = {
+  private val numMatchedRows = longMetric("numMatchedRows")
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:93]: shouldn't this be `lazy`? otherwise `metrics` becomes non lazy...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 14]
```diff
@@ -137,11 +144,20 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
   protected def streamSideKeyGenerator(): UnsafeProjection =
     UnsafeProjection.create(streamedBoundKeys)
 
-  @transient protected[this] lazy val boundCondition = if (condition.isDefined) {
-    Predicate.create(condition.get, streamedPlan.output ++ buildPlan.output).eval _
-  } else {
-    (r: InternalRow) => true
-  }
+  private val numMatchedRows = longMetric("numMatchedRows")
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:147]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 12 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:56]: But do we need to define `metrics` here at all? If `numMatchedRows` becomes `lazy` as it should be, then probably we don't need to have `metrics` here.

Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: I removed the metrics from HashJoin in the latest commit and made the numMatchedRows lazy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 13 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: Done. Fixed in latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 14 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Done. Fixed in latest commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 15]
```diff
@@ -513,12 +527,17 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
...
          |  $conditionPassed = !${ev.isNull} && ${ev.value};
          |}
        """.stripMargin
     } else {
-      s"final boolean $conditionPassed = true;"
+      s"""
+         |if ($matched != null) {
+         |  $numMatched.add(1);
+         |}
+       """.stripMargin + s"final boolean $conditionPassed = true;"
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:540]:
```
      s"""
         |if ($matched != null) {
         |  $numMatched.add(1);
         |}
         |final boolean $conditionPassed = true;
       """.stripMargin
```
?

**Comment by maropu** [comment]
Adding this metric seems fine, but I'm a bit worried if users can understand what `"number of matched rows"` exactly means for each join type. The current user's docs don't have a page to explain about each the SQL metrics in the web UI? I looked around the docs, but I couldn't find them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 13 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:93]: Actually making it lazy is causing a NullPointerException in SqlQuerySuite, ExistenceJoinSuite, OuterJoinSuite.  Somehow the sqlContext object in line 67 is turning out to be null. I am still trying to understand why that is the case. For now I reverted the change.
`java.lang.NullPointerException
	at org.apache.spark.sql.execution.SparkPlan.sparkContext(SparkPlan.scala:67)
	at org.apache.spark.sql.execution.joins.BroadcastNestedLoopJoinExec.metrics$lzycompute(BroadcastNestedLoopJoinExec.scala:42)
	at org.apache.spark.sql.execution.joins.BroadcastNestedLoopJoinExec.metrics(BroadcastNestedLoopJoinExec.scala:41)
	at org.apache.spark.sql.execution.SparkPlan.longMetric(SparkPlan.scala:135)
	at org.apache.spark.sql.execution.joins.BroadcastNestedLoopJoinExec.numMatchedRows$lzycompute(BroadcastNestedLoopJoinExec.scala:93)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 15 above)
Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Done. Fixed in latest commit.

**Comment by sririshindra** [comment]
> Adding this metric seems fine, but I'm a bit worried if users can understand what `"number of matched rows"` exactly means for each join type. The current user's docs don't have a page to explain about each the SQL metrics in the web UI? I looked around the docs, but I couldn't find them.

I agree that it might be confusing to the users. Would you like me to add some documentation here [(monitoring).](https://spark.apache.org/docs/latest/monitoring.html) explaining what the metric does for each join type?

**Comment by maropu** [comment]
> Would you like me to add some documentation here (monitoring). explaining what the metric measures for each join type? cc: @maropu

Yea, I think It would be very helpful if you could do it . Since these metrics are SQL-specific, so it might be better to write them somewhere in [the Spark SQL Guide](https://spark.apache.org/docs/latest/sql-programming-guide.html), then add a link to it in the monitoring page. cc: @gatorsmile @HyukjinKwon

**Comment by sririshindra** [comment]
> > Would you like me to add some documentation here (monitoring). explaining what the metric measures for each join type? cc: @maropu
> 
> Yea, I think It would be very helpful if you could do it . Since these metrics are SQL-specific, so it might be better to write them somewhere in [the Spark SQL Guide](https://spark.apache.org/docs/latest/sql-programming-guide.html), then add a link to it in the monitoring page. cc: @gatorsmile @HyukjinKwon

Ok, I will try to get it done tomorrow.

**Comment by sririshindra** [comment]
@maropu I added some documentation briefly explaining the metric's definition and created a PR in the spark website. Please find the link to that [PR here ](https://github.com/apache/spark-website/pull/317). Please let me know if this is satisfactory and if any more changes are needed.

**Comment by gatorsmile** [comment]
Can we improve the test coverage? Build a new suite instead of modify the existing one.

cc @cloud-fan @maryannxue @rednaxelafx

**Comment by gatorsmile** [comment]
Also cc @viirya I remember we saw the perf regressions when you adding new metrics for Aggregate. Maybe you can also take a look at this

**Comment by sririshindra** [comment]
> Can we improve the test coverage? Build a new suite instead of modify the existing one.
> 
> cc @cloud-fan @maryannxue @rednaxelafx

@gatorsmile Could you please elaborate on what do you mean by a new Suite? Do mean to say that there should be a separate suite for the metric that is being proposed in this PR or are you suggesting that the SQLMetrics Suite itself should be refactored into multiple test Suites (maybe a seperate test Suite for Join based tests, maybe further divide each join based tests with its own test Suite). Also, do you think this new Suite should be done as part of this PR or do you think that a separate follow up PR would be more appropriate?

**Comment by gatorsmile** [comment]
A new suite is preferred here. Creating a new suite for Join-specific SQLMetrics is preferred here, because it has to cover various different cases, including different join types, different join algorithms, whole-stage on/off, and different runtime shortcuts. Also, stream-stream and batch-stream joins should also be tested.

**Comment by maropu** [comment]
@sririshindra could you check if no performance (e.g., TPCDS queries) changes, too?

**Comment by sririshindra** [comment]
> A new suite is preferred here. Creating a new suite for Join-specific SQLMetrics is preferred here, because it has to cover various different cases, including different join types, different join algorithms, whole-stage on/off, and different runtime shortcuts. Also, stream-stream and batch-stream joins should also be tested.

@gatorsmile I refactored the existing test Suite and created a new Suite that contains all the join specifc tests.

**Comment by sririshindra** [comment]
> @sririshindra could you check if no performance (e.g., TPCDS queries) changes, too?

Sure, I am working that right now.

**Comment by bersprockets** [comment]
Retest this please.

**Comment by bersprockets** [comment]
jenkins, retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala [hunk 16]
```diff
@@ -57,7 +57,8 @@ case class BroadcastHashJoinExec(
   }
 
   override lazy val metrics = Map(
-    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+    "numMatchedRows" -> SQLMetrics.createMetric(sparkContext, "number of matched rows"))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: This seems like a "progress bar" instead of a metrics. Most likely people don't care about it because it's usually `num output rows of left` * `num output rows of right`.

Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: I am running the TPC-DS benchmark right now. I will post the results shortly.

Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: I ran the benchmark and I see a clear performance degradation for query-16 and query-94. All the other queries seem to be fine. I will debug why there is a performance degradation for  query-16 & query-94. https://docs.google.com/document/d/16_cjk3g3_7mMH2C_T9bsFroxaH_CihJkXINYirt-mtg/edit?usp=sharing 
cc: @cloud-fan

Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: Update: I figured out why the perf degradation is occurring for query-16 & query-94 and fixed it. I ran the benchmark for a couple of iterations and looks like my fix is working. I am running the benchmark with 25 iterations to identify if any other query has been effected.

Comment (sririshindra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:61]: Update: In my most recent commit I fixed a perf issue that is causing a significant degradation for query-16 & quer-94. However there still seems to be some perf degradation that I cannot fix. Maybe it is because adding the new metric itself or maybe there is something wrong with the way that I am running the tests. I will run the benchmark again on a bare metal machine and see if it makes any difference. I will also see if there is anything else that can be done to improve the performance. My most recent runs are shared in this doc. https://docs.google.com/spreadsheets/d/1A0jCx5BL0wcN28oDQDOs73QhVZ71uNbKDg9Q8S0IE6g/edit?usp=sharing
